### PR TITLE
Power

### PR DESCRIPTION
**Pull request by vikram2000b**

### CODE REVIEW

**Comment by sdavtaker** [comment]
It would be better if you "rebase" in place of "merge" from master for the commits and try to keep all commits atomic. You can do a lot of difference by using rebase before sending a pull request. History edition features can make a big difference later by making "git bisect" usable or not. 
Also, I recommend to read: https://chris.beams.io/posts/git-commit/ for better commit message writing.

File: include/real/real.hpp [hunk 1]
```diff
@@ -638,12 +638,43 @@ namespace boost {
...
              *  @return: returns a new boost real whose value is real_num^exponent
-             *  @author: Kishan Shukla
+             *  @author: Kishan Shukla & Vikram Singh Chundawat
              */
 
             static real power(real<T> real_num, real<T> exponent){
- 
-                return real(real_operation<T>(real_num._real_p, exponent._real_p, OPERATION::POWER));
+
+                //checking whether the number is integer or not
```
Comment (sdavtaker) [inline include/real/real.hpp:646]: nit: You starting with space the other comments.

File: include/real/real.hpp [hunk 2]
```diff
@@ -638,12 +638,43 @@ namespace boost {
...
+                     * Warning: the result of non-integral power of a negative number is a complex number,
+                     * and we do not support complex numbers. So, that will generate error.
+                     **/
+                    
+                    /**
+                     * Now, we will check whether upper bound is negative or not, if it is, then we will throw error
+                     **/
+                    if(itr.get_interval().upper_bound.positive == false){
+                        throw non_integral_power_of_negative_number();
+                    }
```
Comment (sdavtaker) [inline include/real/real.hpp:666]: Why you do not need to test for lowerbound?

File: include/real/real_data.hpp [hunk 3]
```diff
@@ -273,7 +273,7 @@ namespace boost {
 
                     break;
                 }
-                case OPERATION::POWER: {
+                case OPERATION::INTEGER_POWER: {
                     ro.get_rhs_itr().iterate_n_times(ro.get_rhs_itr().maximum_precision());
 
                     if (ro.get_rhs_itr().get_interval().lower_bound != ro.get_rhs_itr().get_interval().upper_bound ||
```
Comment (sdavtaker) [inline include/real/real_data.hpp:279]: We are doing this check many times, maybe it makes sense to make a function in interval to ask this "isPunctual" ?

File: include/real/real_operation.hpp [hunk 4]
```diff
@@ -21,7 +21,7 @@ namespace boost{
         * 
         * @warning due to the recursive nature of real_operation, destruction may cause stack overflow
         */
-        enum class OPERATION{ADDITION, SUBTRACTION, MULTIPLICATION, DIVISION, POWER, EXPONENT, LOGARITHM, SIN, COS, TAN, COT, SEC, COSEC}; 
+        enum class OPERATION{ADDITION, SUBTRACTION, MULTIPLICATION, DIVISION, INTEGER_POWER, EXPONENT, LOGARITHM, SIN, COS, TAN, COT, SEC, COSEC};
```
Comment (sdavtaker) [inline include/real/real_operation.hpp:24]: EXPONENT should be named POWER.

File: test/real_math_test.cpp [hunk 5]
```diff
@@ -72,7 +72,7 @@ TEST_CASE("LOGARITHM AND EXPONENT"){
 		CHECK(b > lower_limit);
 		CHECK(b < upper_limit);
 
-		b = real::exp(a); // ln(0.0047783) = −0.738500258
+		b = real::exp(a); // exo(0.47783) = 1.612571323
```
Comment (sdavtaker) [inline test/real_math_test.cpp:75]: exo->exp

File: include/real/real.hpp (hunk 2 above)
Comment (vikram2000b) [reply include/real/real.hpp:666]: Later checks will be done by logarithmic function(that is why I modified its error message). Doing those checks here will be too cost inefficient, as we may need to iterate until we get to maximum precision, and that will be again done by logarithm. So, I left that task for logarithm function.

File: include/real/real_data.hpp (hunk 3 above)
Comment (vikram2000b) [reply include/real/real_data.hpp:279]: Will do that.

File: include/real/real_operation.hpp (hunk 4 above)
Comment (vikram2000b) [reply include/real/real_operation.hpp:24]: Ok

File: include/real/real.hpp [hunk 6]
```diff
@@ -638,12 +638,44 @@ namespace boost {
...
+             *  @author: Kishan Shukla & Vikram Singh Chundawat
              */
 
-            static real power(real<T> real_num, real<T> exponent){
- 
-                return real(real_operation<T>(real_num._real_p, exponent._real_p, OPERATION::POWER));
+            static real power(real<T> real_num, real<T> power){
+                // checking whether the number is integer or not
+                static real<T> zero("0");
+                static real<T> one("1");
```
Comment (Laouen) [inline include/real/real.hpp:647]: These literals shouldn't be in the literal namespace?

File: include/real/real.hpp [hunk 7]
```diff
@@ -638,12 +638,44 @@ namespace boost {
...
+                    /**
+                     * Now, if number is negative, then logarithm function will check out and throw error
+                     **/
+                    result = real(real_operation<T>(real_num._real_p, zero._real_p, OPERATION::LOGARITHM));
+                    result = real(real_operation<T>(result._real_p, power._real_p, OPERATION::MULTIPLICATION));
+                    result = real(real_operation<T>(result._real_p, zero._real_p, OPERATION::EXPONENT));
+                    return result;
+                }
+                // if power is integer and negative
+                else if(itr.get_interval().lower_bound.positive == false){
```
Comment (Laouen) [inline include/real/real.hpp:672]: a number is negative if its both boundaries are negative, (checking if upper boundary is negative does the job). but the lower boundary as negative gives not mutch information about the number sign.

I think I'm missing something here, but if this is the case, an explanation comment in the code bould be good.

File: include/real/real.hpp (hunk 6 above)
Comment (vikram2000b) [reply include/real/real.hpp:647]: Yes they should be. But we can't use them here. Literals of real class will be defined after real class. And these variables are used inside real class.

File: include/real/real.hpp (hunk 7 above)
Comment (vikram2000b) [reply include/real/real.hpp]: That whole part is now removed. I think I should have checked for upper bound. No specific reason for checking lower bound :) .

**Pull request merged**